### PR TITLE
Support DLPack

### LINKED ISSUE #983: Support DLPack
Recently Facebook Research released [Tensor Comprehension](https://github.com/facebookresearch/TensorComprehensions) which enables to optimize a custom kernel to be as fast as cuDNN operators. This great tool may be able to be integrated with CuPy if CuPy supports DLPack as the GPU array layout. Let's start supporting DLPack in CuPy.

ref: https://github.com/dmlc/dlpack/blob/master/include/dlpack/dlpack.h#L94

### PR DESCRIPTION
**Pull request by mitmul**
This PR resolves #983.

This PR adds a method `toDLPack` to CuPy ndarray. `toDLPack` method returns PyCapsule.
This PR also adds `cupy.fromDLPack` method to convert DLPack tensor into CuPy's ndarray.

### CuPy <-> PyTorch tensor

This enables to convert CuPy arrays to PyTorch tensors, and vice versa.

#### PyTorch -> CuPy
```
import torch
import cupy

from torch.utils.dlpack import to_dlpack
from torch.utils.dlpack import from_dlpack

# ---------------------------------------
# PyTorch -> CuPy
# ---------------------------------------

# Create a PyTorch tensor
tx = torch.randn(1, 2, 3, 4).cuda()
print(tx)

# Convert it into a dlpack tensor
t1 = to_dlpack(tx)

# Convert it further into a CuPy array
cx = cupy.fromDlpack(t1)
print(cx)
```
Output:
```
tensor([[[[-1.0013,  1.1196, -1.0771, -0.7612],
          [ 1.4412, -0.7757,  0.1054,  0.8019],
          [ 0.1014, -0.2111,  0.9335, -0.8446]],

         [[ 0.0085,  0.7669, -1.0250,  3.3684],
          [-0.1365,  0.1029,  0.2827,  0.0307],
          [-1.0236,  1.4164,  0.9975,  0.8236]]]], device='cuda:0')
[[[[-1.0012635   1.11960101 -1.07713795 -0.76120251]
   [ 1.4411695  -0.77568114  0.1054479   0.80189544]
   [ 0.10137261 -0.21112843  0.93348098 -0.84461737]]

  [[ 0.0085327   0.76691639 -1.02497101  3.36835527]
   [-0.13651586  0.10294367  0.28265193  0.03072196]
   [-1.02356339  1.41642499  0.99751163  0.82362241]]]]
```

#### CuPy -> PyTorch

```
# --------------------------------------
# CuPy -> PyTorch
# --------------------------------------

ca = cupy.random.randn(1, 2, 3, 4).astype(cupy.float32)
print(ca)

t2 = ca.toDlpack()

cb = from_dlpack(t2)
print(cb)
```
Output:
```
[[[[-1.85036588  0.40208972  1.1292758  -0.7116102 ]
   [ 0.16188958  0.39067528  2.11596918  0.27136478]
   [-0.6051448   0.58887625 -1.22710478 -0.90009344]]

  [[-1.31603026 -0.30327141 -0.28411245  0.32678974]
   [-1.0101825   0.19829388 -0.35092002  0.24203797]
   [-0.30412203  2.31131864  1.43144453  0.45414579]]]]
tensor([[[[-1.8504,  0.4021,  1.1293, -0.7116],
          [ 0.1619,  0.3907,  2.1160,  0.2714],
          [-0.6051,  0.5889, -1.2271, -0.9001]],

         [[-1.3160, -0.3033, -0.2841,  0.3268],
          [-1.0102,  0.1983, -0.3509,  0.2420],
          [-0.3041,  2.3113,  1.4314,  0.4541]]]], device='cuda:0')
```

### TODO

- [x] Implement `toDLPack` method
- [x] Implement `fromDLPack` method
- [x] Add tests
- [x] Add documents
- [x] Add comments (why we use `toDlpack` and `fromDlpack` for the names of the functions: ref. https://facebookresearch.github.io/TensorComprehensions/integrating_any_ml_framework.html)

### Discussion

- ~Who is responsible for managing the memory region of a DLTensor after we use `fromDLPack` to make a CuPy ndarray from it? The DLTensor might be created by another framework. (ref: https://github.com/dmlc/dlpack/issues/18)~

### Reference

- Implementation in ATen: https://github.com/zdevito/ATen/blob/master/src/ATen/DLConvertor.cpp

### CODE REVIEW

**Comment by mitmul** [comment]
Actually I followed this documentation: [Integrating TC with ML framework](https://facebookresearch.github.io/TensorComprehensions/integrating_any_ml_framework.html). It suggests to use `toDlpack` for the converting method (I noticed it should be `toDlpack`, not `toDLPack`... I'll fix the name). So, some users may assume CuPy has a method with that name, so I think it's better to follow the suggestion found in the TC documentation.

**Comment by beam2d** [comment]
Ah, I see. OK, I understood the background. Maybe that should be noted in a comment somewhere so that we will not accidentally rename it.

File: cupy/core/dlpack.pyx [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    dl_tensor.ndim = ndim
+    dl_tensor.dtype = dereference(dtype)
+
+    cdef int64_t* shape = <int64_t*>malloc(ndim * sizeof(int64_t))
+    for n in range(ndim):
+        shape[n] = array._shape[n]
+    dl_tensor.shape = shape
+
+    cdef int n_strides = len(array._strides)
+    cdef int64_t* strides = <int64_t*>malloc(n_strides * sizeof(int64_t))
```
Comment (okuta) [inline cupy/core/dlpack.pyx]: Who does free this pointer?

File: cupy/core/dlpack.pyx [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    elif cupy.issubdtype(array.dtype, cupy.integer):
+        dtype.code = <uint8_t>DLDataTypeCode.kDLInt
+    elif cupy.issubdtype(array.dtype, cupy.floating):
+        dtype.code = <uint8_t>DLDataTypeCode.kDLFloat
+    else:
+        raise ValueError('Unknown dtype')
+    dtype.bits = <uint8_t>(array.dtype.itemsize * 8)
+    dtype.lanes = <uint16_t>1
+
+    cdef DLTensor* dl_tensor = <DLTensor*>malloc(sizeof(DLTensor))
```
Comment (okuta) [inline cupy/core/dlpack.pyx]: Who does free this pointer?

File: cupy/core/dlpack.pyx [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+    cdef int n_strides = len(array._strides)
+    cdef int64_t* strides = <int64_t*>malloc(n_strides * sizeof(int64_t))
+    for n in range(n_strides):
+        strides[n] = array._strides[n]
+    dl_tensor.strides = strides
+
+    dl_tensor.byte_offset = 0
+
+    cdef DLManagedTensor* dlm_tensor = \
```
Comment (okuta) [inline cupy/core/dlpack.pyx]: Who does free this pointer?

File: cupy/core/dlpack.pyx [hunk 4]
```diff
@@ -0,0 +1,57 @@
+from libc.stdlib cimport malloc
+from cython.operator cimport dereference
+
+import cupy
+
+from cupy.core.core cimport ndarray
```
Comment (okuta) [inline cupy/core/dlpack.pyx]: Could you try to use `from cupy.core cimport core` . Perhaps we cannot ...

File: cupy/core/dlpack.pyx [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+from cupy.cuda.runtime cimport free
```
Comment (okuta) [inline cupy/core/dlpack.pyx]: ditto

File: cupy/core/dlpack.pyx [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    for n in range(n_strides):
+        strides[n] = array._strides[n]
+    dl_tensor.strides = strides
+
+    dl_tensor.byte_offset = 0
+
+    cdef DLManagedTensor* dlm_tensor = \
+        <DLManagedTensor*>malloc(sizeof(DLManagedTensor))
+    dlm_tensor.dl_tensor = dereference(dl_tensor)
+    dlm_tensor.manager_ctx = array.get_pointer().ptr
```
Comment (okuta) [inline cupy/core/dlpack.pyx]: Please use `array.data.ptr`. It is faster.

File: cupy/core/dlpack.pyx [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+            raise TypeError('float{} is not supported.'.format(dtype.bits))
+    else:
+        raise TypeError('Unsupported dtype. dtype code: {}'.format(dtype.code))
+
+    mem_ptr = memory.MemoryPointer(mem, mem.dlm_tensor.dl_tensor.byte_offset)
+    cdef int64_t ndim = mem.dlm_tensor.dl_tensor.ndim
+    cdef int64_t* shape = mem.dlm_tensor.dl_tensor.shape
+    cdef int64_t* strides = mem.dlm_tensor.dl_tensor.strides
+    for i in range(ndim):
+        strides[i] = strides[i] * (dtype.bits // 8)
```
Comment (okuta) [inline cupy/core/dlpack.pyx]: Please do not edit original data.

Comment (mitmul) [reply cupy/core/dlpack.pyx]: Right. Fixed.

File: cupy/core/dlpack.pyx [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+    cdef object dltensor
+
+    def __init__(self, object dltensor):
+        self.dltensor = dltensor
+        self.dlm_tensor = <DLManagedTensor *>pycapsule.PyCapsule_GetPointer(
+            dltensor, 'dltensor')
+        self.device = cupy.cuda.Device(self.dlm_tensor.dl_tensor.ctx.device_id)
+        self.ptr = self.dlm_tensor.dl_tensor.data
+        cdef int n = 0
+        for s in self.dlm_tensor.dl_tensor.shape[
```
[the remark this answers is not in the document]
Comment (mitmul) [reply cupy/core/dlpack.pyx]: OK, fixed to:
```
        cdef int ndim = self.dlm_tensor.dl_tensor.ndim
        cdef int64_t* shape = self.dlm_tensor.dl_tensor.shape[:ndim]
        for s in shape:
            n += s
```

File: cupy/__init__.py [hunk 9]
```diff
@@ -598,6 +598,12 @@ def isscalar(num):
 from cupy.core import ElementwiseKernel  # NOQA
 from cupy.core import ReductionKernel  # NOQA
 
+# -----------------------------------------------------------------------------
+# DLPack
+# -----------------------------------------------------------------------------
+
+from cupy.core import fromDlpack  # NOQA
+from cupy.core import toDlpack  # NOQA
```
[the remark this answers is not in the document]
Comment (mitmul) [reply cupy/__init__.py]: Right. deleted

**Comment by mitmul** [comment]
@okuta Thanks! I modified a bit to avoid this error:
```
ValueError: PyCapsule_GetPointer called with incorrect name
```
Using your test code as is causes this error because `pycapsule_deleter` retrieves the pointer of DLManagedTensor with the name of `used_dltensor`. So we can't delete an unused dltensor despite it allocates some memory.

This is the fixed test case:
```

class TestDLTensorMemory(unittest.TestCase):

    def test_deleter(self):
        pool = cupy.get_default_memory_pool()
        pool.free_all_blocks()
        array = cupy.empty(10)
        tensor = array.toDlpack()
        assert pool.n_free_blocks() == 0
        del array
        assert pool.n_free_blocks() == 0
        reverted_array = cupy.fromDlpack(tensor)
        del reverted_array
        del tensor
        assert pool.n_free_blocks() == 1
```

Another way to avoid the above error is to update `pycapsule_deleter` from
```
cdef void pycapsule_deleter(object dltensor):
    cdef DLManagedTensor* dlm_tensor = \
        <DLManagedTensor *>pycapsule.PyCapsule_GetPointer(
            dltensor, 'used_dltensor')
    deleter(dlm_tensor)
```
to
```
cdef void pycapsule_deleter(object dltensor):
    cdef DLManagedTensor* dlm_tensor
    try:
        dlm_tensor = <DLManagedTensor *>pycapsule.PyCapsule_GetPointer(
            dltensor, 'used_dltensor')
    except:
        dlm_tensor = <DLManagedTensor *>pycapsule.PyCapsule_GetPointer(
            dltensor, 'dltensor')
    deleter(dlm_tensor)
```

**Comment by mitmul** [comment]
Sorry, I fixed this again by choosing the second option (modyfing `pycapsule_deleter` function). So now your test code is running as is.

**Comment by tqchen** [comment]
Glad to see this PR, any chance you might want add TVM support? Since DLPack is already supported, adding TVM support can be done out of box by registering the Tensor as extension type, see https://github.com/dmlc/tvm/issues/658

File: cupy/core/dlpack.pyx [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+        <capsule object "dltensor" at ...>
+        >>> array2 = cupy.fromDlpack(dltensor)
+        >>> cupy.testing.assert_array_equal(array1, array2)
+
+    """
+    mem = DLPackMemory(dltensor)
+
+    cdef DLDataType dtype = mem.dlm_tensor.dl_tensor.dtype
+    if dtype.code == DLDataTypeCode.kDLUInt:
+        if dtype.bits == 8:
```
Comment (okuta) [inline cupy/core/dlpack.pyx]: Please use local variable `cdef int bits = dtype.bits` .

**Comment by mitmul** [comment]
@tqchen Thank you for commenting on this PR. Sorry for late reply. TVM support sounds great to me, so let me bring up it for discussion in internal roadmap meeting.

File: cupy/core/dlpack.pyx [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+            raise TypeError('int{} is not supported.'.format(bits))
+    elif dtype.code == DLDataTypeCode.kDLFloat:
+        if bits == 16:
+            cp_dtype = cupy.float16
+        elif bits == 32:
+            cp_dtype = cupy.float32
+        elif bits == 64:
+            cp_dtype = cupy.float64
+        else:
+            raise TypeError('float{} is not supported.'.format(dtype.bits))
```
Comment (okuta) [inline cupy/core/dlpack.pyx]: You can use `bits`

File: cupy/core/dlpack.pyx [hunk 12]
```diff
@@ -0,0 +1,250 @@
...
+    cdef int64_t ndim = mem.dlm_tensor.dl_tensor.ndim
+
+    cdef int64_t* shape = mem.dlm_tensor.dl_tensor.shape
+    cdef vector[Py_ssize_t] shape_vec
+    shape_vec.assign(shape, shape + ndim)
+
+    cdef int64_t* strides = mem.dlm_tensor.dl_tensor.strides
+    cdef vector[Py_ssize_t] strides_vec
+    for i in range(ndim):
+        strides_vec.push_back(strides[i] * (dtype.bits // 8))
```
Comment (okuta) [inline cupy/core/dlpack.pyx]: You can use `bits`

File: cupy/core/dlpack.pyx (hunk 11 above)
Comment (mitmul) [reply cupy/core/dlpack.pyx]: Oops, thanks

File: cupy/core/dlpack.pyx (hunk 12 above)
Comment (mitmul) [reply cupy/core/dlpack.pyx]: Yes, fixed

**Comment by mitmul** [comment]
Jenkins test this please

**Comment by okuta** [comment]
LGTM!

**Pull request merged**